pkg/networkservice/chains/client/client.go

```diff
@@ -21,146 +19,53 @@ package client
-//             - cc - grpc.ClientConnInterface for the endpoint to which this client should connect
-func NewClient(ctx context.Context, connectTo *url.URL, clientOpts ...Option) networkservice.NetworkServiceClient {
-	rv := new(networkservice.NetworkServiceClient)
+func NewClient(ctx context.Context, clientOpts ...Option) networkservice.NetworkServiceClient {
 	var opts = &clientOptions{
 		name:            "client-" + uuid.New().String(),
 		authorizeClient: null.NewClient(),
-		dialTimeout:     100 * time.Millisecond,
+		refreshClient:   refresh.NewClient(ctx),
 	}
 	for _, opt := range clientOpts {
 		opt(opts)
 	}
 
-	*rv = chain.NewNetworkServiceClient(
-		updatepath.NewClient(opts.name),
-		begin.NewClient(),
-		metadata.NewClient(),
-		refresh.NewClient(ctx),
-		adapters.NewServerToClient(
-			chain.NewNetworkServiceServer(
-				heal.NewServer(ctx,
-					heal.WithOnHeal(rv),
-					heal.WithOnRestore(heal.OnRestoreRestore),
-					heal.WithRestoreTimeout(time.Minute)),
-				clienturl.NewServer(connectTo),
-				connect.NewServer(ctx, func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
-					return chain.NewNetworkServiceClient(
-						append(
-							opts.additionalFunctionality,
-							// TODO: move back to the end of the chain when `begin` chain element will be ready
-							heal.NewClient(ctx, networkservice.NewMonitorConnectionClient(cc),
-								heal.WithEndpointChange()),
-							opts.authorizeClient,
-							networkservice.NewNetworkServiceClient(cc),
-						)...,
-					)
-				},
-					connect.WithDialOptions(opts.dialOptions...),
-					connect.WithDialTimeout(opts.dialTimeout)),
-			),
-		),
-	)
-	return *rv
-}
-
-// NewClientFactory - returns a (3.) case func(cc grpc.ClientConnInterface) NSM client factory.
-func NewClientFactory(clientOpts ...Option) connect.ClientFactory {
-	return func(ctx context.Context, cc grpc.ClientConnInterface) networkservice.NetworkServiceClient {
-		var rv networkservice.NetworkServiceClient
-		var opts = &clientOptions{
-			name:            "client-" + uuid.New().String(),
-			authorizeClient: null.NewClient(),
-		}
-		for _, opt := range clientOpts {
-			opt(opts)
-		}
-		rv = chain.NewNetworkServiceClient(
+	return chain.NewNetworkServiceClient(
+		append(
+			[]networkservice.NetworkServiceClient{
+				updatepath.NewClient(opts.name),
+				begin.NewClient(),
+				metadata.NewClient(),
+				opts.refreshClient,
+				clienturl.NewClient(opts.clientURL),
+				clientconn.NewClient(opts.cc),
+			},
 			append(
-				append([]networkservice.NetworkServiceClient{
-					updatepath.NewClient(opts.name),
-					begin.NewClient(),
-					metadata.NewClient(),
-					refresh.NewClient(ctx),
-					// TODO: move back to the end of the chain when `begin` chain element will be ready
-					heal.NewClient(ctx, networkservice.NewMonitorConnectionClient(cc)),
-				}, opts.additionalFunctionality...),
+				opts.additionalFunctionality,
+				dial.NewClient(ctx,
```

Comment:
Probably we may want to put `dial.NewClient()` before `opts.additionalFunctionality`?
In such case we can build client that can access some other gRPC services on the remote side (e.g. monitor or some user-defined service) or check connection status for its own purposes.

Reply:
Client takes connection from metadata, so it can access it on return - no need for changing element order.